lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -3075,8 +3080,18 @@ SILLocation::FilenameAndLocation
 IRGenDebugInfoImpl::decodeSourceLoc(SourceLoc SL) {
   auto &Cached = FilenameAndLocationCache[SL.getOpaquePointerValue()];
   if (Cached.filename.empty()) {
-    Cached = sanitizeCodeViewFilenameAndLocation(
-        SILLocation::decode(SL, SM, /*ForceGeneratedSourceToDisk=*/true));
+    Cached = sanitizeCodeViewFilenameAndLocation(SILLocation::decode(
+        SL, SM, /*MacroExpansionOpts=*/{true, Opts.MacroExpansionFilesPath}));
+
+    // If `-debug-prefix-map` would change the path, handle that here.
+    // This is necessary to handle macro expansions correctly.
+    if (!Opts.MacroExpansionFilesPath.empty() &&
+        Cached.filename.startswith(Opts.MacroExpansionFilesPath)) {
```

Comment:
I'm not crazy about this check, but without it (remapping the path unconditionally), I end up with an extra `DIFile` entry with the wrong path in `test/DebugInfo/debug_prefix_map_abs_rel_subdir.swift`:

```
# Correct
!14 = !DIFile(filename: "debug_prefix_map_abs_rel_subdir.swift", directory: "./SubdirectoryWithAVeryUniqueName")
# ... then later, this incorrect one ...
!51 = !DIFile(filename: "SubdirectoryWithAVeryUniqueName/debug_prefix_map_abs_rel_subdir.swift", directory: "./SubdirectoryWithAVeryUniqueName")
```

I don't entirely understand the logic here but this "worked", at least for these tests. I'd like to have a better solution.